src/types/osstr.rs

```diff
@@ -0,0 +1,191 @@
+            // no methods to get a view into its internal buffers, so this step can only be done via
+            // copying.
+            // Second, we have to convert the wide char buffer into a OsString. Rust provides no
+            // way to cheaply convert an owned buffer into an OsString; it only exposes a method
+            // involving a copy and an allocation.
+
+            // Get an owned allocated wide char buffer from PyString, which we have to deallocate
+            // ourselves
+            let mut size: ffi::Py_ssize_t = 0;
+            let buffer_ptr =
+                unsafe { ffi::PyUnicode_AsWideCharString(pystring.as_ptr(), &mut size) };
+
+            // Copy wide char buffer into OsString
+            let os_string = std::os::windows::ffi::OsStringExt::from_wide(unsafe {
+                std::slice::from_raw_parts(buffer_ptr, size as usize)
+            });
+
+            unsafe {
+                ffi::PyMem_Free(buffer_ptr as *mut libc::c_void);
+            }
+
+            Ok(os_string)
+        }
+    }
+}
+
+impl<'a> IntoPy<PyObject> for &'a OsStr {
+    #[inline]
+    fn into_py(self, py: Python) -> PyObject {
+        self.to_object(py)
+    }
+}
+
+impl<'a> ToPyObject for Cow<'a, OsStr> {
+    #[inline]
+    fn to_object(&self, py: Python) -> PyObject {
+        (&self as &OsStr).to_object(py)
+    }
+}
+
+impl ToPyObject for OsString {
+    #[inline]
+    fn to_object(&self, py: Python) -> PyObject {
+        (&self as &OsStr).to_object(py)
+    }
+}
+
+impl IntoPy<PyObject> for OsString {
+    fn into_py(self, py: Python) -> PyObject {
+        self.to_object(py)
+    }
+}
+
+impl<'a> IntoPy<PyObject> for &'a OsString {
+    #[inline]
+    fn into_py(self, py: Python) -> PyObject {
+        self.to_object(py)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::{IntoPy, Python};
+    use std::ffi::{OsStr, OsString};
+
+    #[test]
+    #[cfg(not(windows))]
+    fn test_non_utf8_conversion() {
+        Python::with_gil(|py| {
+            use std::os::unix::ffi::OsStrExt;
+
+            // this is not valid UTF-8
+            let payload = &[250, 251, 252, 253, 254, 255, 0, 255];
+            let os_str = OsStr::from_bytes(payload);
+
+            // do a roundtrip into Pythonland and back and compare
+            let py_str = os_str.into_py(py);
+            let os_str_2: &OsStr = py_str.extract(py).unwrap();
+            assert_eq!(os_str, os_str_2);
+        });
+    }
+
+    #[test]
+    fn test_conversion() {
+        Python::with_gil(|py| {
+            let os_str = OsStr::new("Hello\0\n🐍");
+
+            // do a roundtrip into Pythonland and back and compare
+            let py_str = os_str.into_py(py);
+            let os_str_2: OsString = py_str.extract(py).unwrap();
+            assert_eq!(os_str, os_str_2);
+        })
+    }
+
+    // HELP: Are these tests enough? I was unsure how to best test it, because of the platform-
+    // dependant nature of OsStr/OsString
```

Comment:
We recently used `proptest` to test roundtripping of u128 / i128 - would be a great fit here too as it'll help us sniff out any edge cases where we have issues.